qiskit/qpy/interface.py

```diff
@@ -153,13 +170,22 @@ def dump(
     else:
         raise TypeError(f"'{program_type}' is not supported data type.")
 
+    if version is None:
+        version = common.QPY_VERSION
+    elif version not in {common.QPY_COMPATIBILITY_VERSION, common.QPY_VERSION}:
+        raise ValueError(
+            f"The specified QPY version {version} is not support for dumping with this version, "
+            f"of Qiskit. The only supported versions are {common.QPY_COMPATIBILITY_VERSION} and "
+            f"{common.QPY_VERSION}"
+        )
+
     version_match = VERSION_PATTERN_REGEX.search(__version__)
     version_parts = [int(x) for x in version_match.group("release").split(".")]
     encoding = type_keys.SymExprEncoding.assign(use_symengine)
     header = struct.pack(
         formats.FILE_HEADER_V10_PACK,
         b"QISKIT",
-        common.QPY_VERSION,
+        version,
```

Comment:
This is an exciting direction to go in. Since qpy currently only supports dumping one version, I was expecting to see more code infrastructure here. I guess in the interest of time with 1.0 approaching this PR just includes the API  for choosing the version but the infrastructure for dumping multiple versions and testing those versions will come when the next qpy change is needed?

Would this change allow qiskit-ibm-runtime to stop vendoring qpy code? I could see the provider API reporting which qpy versions it supported and then the provider client selecting the maximum version that the installed qiskit supported. Since qpy versions are linear, I guess the provider just needs to report the maximum version and then the client would use the qpy version if it could and the compatibility version otherwise. The ability to dump only compatibility and latest might be a problem here -- if you are using a dev version of qiskit with qpy 12 but the provider still has qpy 11, the dev client would only have 10 and 12 to choose from and would have to fall back to 10 when perhaps 11 added the ability to serialize some new class that the user wanted to use.

Reply:
The present implementation is making a couple of tacit assumptions that there's only actually one possible version right now, despite this line - there's no actual behaviour switching anywhere.

The aim is to get the flag in place for 1.0 so that in the future when we _do_ add a new QPY version, the pieces are already in places.

Imo (and this what I'd understood the intent to be, but maybe I'm mistaken) we should support exporting to _any_ QPY version >= the minimum compat version.  So if min compat is 10 and we've released both 11 and 12, we should accept 10, 11 and 12 in this argument, which I think alleviates your concern about the dev version.

Reply:
The release note says

>    This is useful if you need to generate a QPY
>    file that will be loaded by an older version of Qiskit. However, the
>     supported versions to emit are very limited, only the latest QPY version
>    (which is the default), and the compatibility QPY version
>    which is :ref:`qpy_version_10` (which was introduced in Qiskit 0.45.0).

and the docstring says

> for compatibility reasons if you need to load the generated QPY payload with an older version of Qiskit you can also select the minimum supported export version

so I was taking that as meaning you couldn't take versions in between, but maybe it doesn't rule that out since intermediate versions do not yet exist. The other thing that made me take this interpretation is the test

    elif version not in {common.QPY_COMPATIBILITY_VERSION, common.QPY_VERSION}:

which could be

    elif version < common.QPY_COMPATIBILITY_VERSION or version > common.QPY_VERSION:

if the intent was to support more than two.

I haven't thought about how much overhead it would take to support 3 or 4 versions versus 2.

Reply:
Oh, I guess the model I had in my head from prior discussion was not exactly what Matt had written in the PR, and I'd kind of overlooked it in the fog of "everything's only actually one version at the moment".

Imo if we're doing this, we have to support export to every version between the minimum and current, and the minimum can't be raised until a major.  Anything else is against the stability guarantees of 1.0, imo.

From what I've seen of how QPY's evolved so far, I don't think that's going to be a hugely onerous ask, provided we leave ourselves the bail out of "we will not backport bugfixes to previous QPY versions".

Reply:
My intent was just the two versions the min and max. I was trying to limit the commitment here to minimize the dev/maintenance overhead by doing this. I am concerned about committing to supporting every QPY version version between the min and max.

My hesitation is mostly around some of the changes we made in the earlier qpy versions which are very different from each other and would basically have meant keeping two parallel serialization paths alive at the same time.  I'm also mildly worried about the overhead for non-format related bugs where we have to fix the same bug potentially in multiple code paths. But I'm not super tied to this and if you both think we can manage supporting ~5 versions (assuming we have a version 11 in 1.0 too and we raise the compat version in 2.0) we can give it a try. I do think it's easy to expand the supported exporter versions in a 1.x release but probably impossible to walk it back (even if we could change it in 2.0, I expect users will expect it).

Reply:
Thinking about this a bit more I think the use case for enabling version negotiation (up to the min compatibility version) probably outweighs my concerns about maintenance. I'll revise the docstring and valid version check accordingly to reflect this. I think we'll probably need to revise the writer infrastructure in 1.1 to better support this model.

Reply:
@mtreinish Yes, I think that is the tradeoff that I was not sure about. If Qiskit can output the intermediate versions, it enables the user to use the latest Qiskit and still work with a remote service that has a relatively recent version of qpy, which I think is a nice enablement. On the other hand, it does require extra maintenance burden for qpy. I don't think the format changes too much now besides adding new things? Maybe there is a way to flag writer methods with qpy versions so that the code doesn't get too unwieldy.

Reply:
Personally I was kind of baking in a refactor of the QPY writer code into supporting this - I think the abstractions we've got in place aren't as helpful as they could be, and others in practice have led to hard to follow code.

Reply:
I adjusted the docs and version check in: https://github.com/Qiskit/qiskit/pull/11644/commits/3350dee78b296eb55a8551f1965c81da8375bdf4

Reply:
Nice, that looks go to me. I hope it's not too much maintenance burden. I wonder if the text needs to say 1.x.y, but it's fine as long as someone remembers to change it with 2.0 :slightly_smiling_face: